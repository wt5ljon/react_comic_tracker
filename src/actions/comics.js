import uuid from 'uuid';
import database from '../firebase/firebase';

export const addComic = (comic) => {
  return {
    type: 'ADD_COMIC',
    comic
  };
};

export const startAddComic = (comicData = {}) => {
  return (dispatch) => {
    const {
      seriesName = '',
      seriesNumber = undefined,
      storyBy = '',
      artBy = '',
      publicationDate = undefined,
      readStatus = false
    } = comicData;
  const comic = { seriesName, seriesNumber, storyBy, artBy, publicationDate, readStatus };
  database.ref('comics').push(comic).then((ref) => {
    dispatch(addComic({
      id: ref.key,
      ...comic
    }))
  })
  };
};

export const removeComic = (id) => ({
  type: 'REMOVE_COMIC',
  id
});

export const editComic = (id, editedComic) => ({
  type: 'EDIT_COMIC',
  id,
  editedComic
});

export const toggleReadStatus = (id) => ({
  type: 'TOGGLE_READ_STATUS',
  id  
});

export const setComics = (comics) => ({
  type: 'SET_COMICS',
  comics
});

export const startSetComics = () => {
  return (dispatch) => {
    return database.ref('comics')
      .once('value')
      .then((snapshot) => {
        const comics = [];
        snapshot.forEach((childSnapshot) => {
          comics.push({
            id: childSnapshot.key,
            ...childSnapshot.val()
          });
        });
        dispatch(setComics(comics));
      });
  };
};
