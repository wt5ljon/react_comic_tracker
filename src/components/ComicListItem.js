import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import moment from 'moment';

const ComicListItem = ({
  id, seriesName, seriesNumber, publicationDate, storyBy, artBy
}) => (
  <Link className="list-item" to={`/edit/${id}`}>
    <div>
      <h3 className="list-item__title">{seriesName}</h3>
      <span className="list-item__sub-title">
        {'Publication Date: '}
        {moment(publicationDate).format('MMMM Do, YYYY')}
      </span>
      <span className="list-item__sub-title">
        {'Story By: '}
        {storyBy}
      </span>
      <span className="list-item__sub-title">
        {'Art By: '}
        {artBy}
      </span>
    </div>
    <h3 className="list-item__data">{seriesNumber}</h3>
  </Link>
);

ComicListItem.propTypes = {
  id: PropTypes.string.isRequired,
  seriesName: PropTypes.string.isRequired,
  seriesNumber: PropTypes.number.isRequired,
  storyBy: PropTypes.string.isRequired,
  artBy: PropTypes.string.isRequired,
  publicationDate: PropTypes.number.isRequired
};

export default ComicListItem;
