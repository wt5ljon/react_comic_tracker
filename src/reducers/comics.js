const initialState = [];

const comicsReducer = (state = initialState, action) => {
  switch (action.type) {
    case 'ADD_COMIC':
      return [
        ...state,
        action.comic
      ];
    case 'REMOVE_COMIC':
      return state.filter((comic) => comic.id !== action.id);
    case 'EDIT_COMIC':
      return state.map((comic) => {
        if (comic.id === action.id) {
          return {
            ...comic,
            ...action.editedComic
          };
        } else {
          return comic;
        }
      });
    case 'SET_COMICS':
      return action.comics;
    default:
      return state;
  };
};

export default comicsReducer;