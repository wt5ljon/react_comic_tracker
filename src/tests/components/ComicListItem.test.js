import React from 'react';
import { shallow } from 'enzyme';
import ComicListItem from '../../components/ComicListItem';
import comics from '../fixtures/comics';

test('should render comic list item', () => {
  const wrapper = shallow(<ComicListItem {...comics[1]} />);
  expect(wrapper).toMatchSnapshot();
});