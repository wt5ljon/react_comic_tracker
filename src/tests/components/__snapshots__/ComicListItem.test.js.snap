// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render comic list item 1`] = `
<Link
  className="list-item"
  replace={false}
  to="/edit/2"
>
  <div>
    <h3
      className="list-item__title"
    >
      Fantastic Four
    </h3>
    <span
      className="list-item__sub-title"
    >
      Publication Date: 
      December 27th, 1969
    </span>
    <span
      className="list-item__sub-title"
    >
      Story By: 
      Lee
    </span>
    <span
      className="list-item__sub-title"
    >
      Art By: 
      Ditko
    </span>
  </div>
  <h3
    className="list-item__data"
  >
    13
  </h3>
</Link>
`;
